granule.py

```diff
@@ -0,0 +1,391 @@
+"""
+This module contains functions that will segment images for 
+granule detection.
+"""
+
+import os
+import matplotlib.pyplot as plt
+import matplotlib.lines as lines
+import matplotlib.patheffects as mpe
+import matplotlib.colors as colors
+import numpy as np
+import scipy.ndimage as sndi
+import skimage
+import sunpy
+import sunpy.map
+from sklearn.cluster import KMeans as KMeans
+
+__all__ = [
+    "segment",
+    "get_threshold",
+    "trim_intergranules",
+    "mark_faculae",
+    "segment_kmeans",
+    "cross_correlation",
+]
+
+def segment(id, data_map, skimage_method, res, plot_intermed=True,
+            mark_dim_centers=False, out_dir='output/'):
+
+    """
+    Segment optical image of the solar photosphere into tri-value maps
+    with 0 = intergranule, 0.5 = faculae, 1 = granule.
+    ----------
+    Parameters:
+        id (string): identifying name of map to be segmented, for file naming
+        data_map (SunPy map): SunPy map containing data to segment
+        skimage_method (string): skimage thresholding method -
+                                options are 'otsu', 'li', 'isodata',
+                                'mean', 'minimum', 'yen', 'triangle'
+        plot_intermed (True or False): whether to plot and save intermediate
+                                data product image
+        mark_dim_centers (True or False): whether to mark dim granule centers
+                                as a seperate catagory for future exploration
+        out_dir (str): Desired directory in which to save intermediate data
+                                product image (if plot_intermed = True);
+        res (float): Spatial resolution (arcsec/pixel) of the data
+    ----------
+    Returns:
+        segmented_map (SunPy map): SunPy map containing segmentated image
+                              (with the original header)
+    """
+
+    if type(data_map) != sunpy.map.mapbase.GenericMap:
+        raise TypeError('Input must be sunpy map.')
+
+    methods = ['otsu', 'li', 'isodata', 'mean', 'minimum', 'yen', 'triangle']
+    if skimage_method not in methods:
+        raise TypeError('Method must be one of: ' + str(methods))
+
+    data = data_map.data
+    header = data_map.meta
+
+    # apply median filter
+    median_filtered = sndi.median_filter(data, size=3)
+
+    # apply threshold
+    threshold = get_threshold(median_filtered, skimage_method)
+
+    # initial skimage segmentation
+    segmented_image = np.uint8(median_filtered > threshold)
+
+    # fix the extra IGM bits in the middle of granules
+    segmented_image_fixed = trim_intergranules(segmented_image,
+                                               mark=mark_dim_centers)
+
+    # mark faculae
+    segmented_image_markfac = mark_faculae(segmented_image_fixed, data, res)
+
+    if plot_intermed:

```

Comment:
I am not in favour of the function being able to plot.

What was the original use case for this?

Reply:
Hmm, I think on the surface. Telling a user if they want to, what each step is doing is useful.
At least from a debugging point of view.

Is there values or information that could be printed to a console that would be useful instead of plotting?

Reply:
Perhaps I could replace this with something such as printing the number of granules and faculae found?

Reply:
Let's start with that. 

I do wonder if that plotting code is more useful in an example, I don't want it to be forgotten.